### PR TITLE
feat(ecs): add `maxSwap` and `swappiness` properties to LinuxParameters

### LINKED ISSUE #18460: aws-ecs: Support MaxSwap and Swappiness in LinuxParameters
### Description

Please support MaxSwap (https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/cfnspec/spec-source/specification/000_cfn/000_official/000_AWS_ECS.json#L975) and Swappiness (https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/cfnspec/spec-source/specification/000_cfn/000_official/000_AWS_ECS.json#L987) in LinuxParameters (https://github.com/aws/aws-cdk/blob/master/packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts).

### Use Case

Using swap allows the container to write excess memory requirements to disk when the container has exhausted all the RAM that is available to it.

### Proposed Solution
```
export interface LinuxParametersProps {
  /**
   * Specifies whether to run an init process inside the container that forwards signals and reaps processes.
   *
   * @default false
   */
  readonly initProcessEnabled?: boolean;

  /**
   * The value for the size (in MiB) of the /dev/shm volume.
   *
   * @default No shared memory.
   */
  readonly sharedMemorySize?: number;

  /**
   * The total amount of swap memory (in MiB) a container can use.
   *
   * @default No swap.
   */
  readonly maxSwap?: number;

  /**
   * This allows you to tune a container's memory swappiness behavior.
   *
   * @default 60
   */
  readonly swappiness?: number;
}

constructor(scope: Construct, id: string, props: LinuxParametersProps = {}) {
  super(scope, id);

  this.sharedMemorySize = props.sharedMemorySize;
  this.initProcessEnabled = props.initProcessEnabled;
  this.maxSwap = props.maxSwap;
  this.swappiness = props.swappiness;
}
```
### Other information

_No response_

### Acknowledge

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

### PR DESCRIPTION
**Pull request by flavioleggio**
Add support to `MaxSwap ` and `Swappiness` attributes in the `LinuxParameters` construct.

Closes #18460

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

File: packages/@aws-cdk/aws-ecs/test/container-definition.test.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/test/container-definition.test.ts:1761]: All of these tests need to verify that the correct values are in the resulting template in the THEN block of the test.

File: packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts:39]: Please include the min and max values in the description. As well as a short description of what 'swappiness' means. Feel free to lift from the ECS docs: https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/aws-properties-ecs-taskdefinition-linuxparameters.html

File: packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts:94]: We should validate these values. Validation logic can either be in this constructor, or in the TaskDefinition constructor when the LinuxParameters is actually used.

**Review by madeline-k — CHANGES_REQUESTED** [review]
Thanks for making this contribution, @flavioleggio! 

Some requests:

1. Please add validation somewhere for the restrictions of maxSwap and swappiness - e.g.  min/max values, if swappiness is supplied but maxSwap is not then swapping won't occur, these values can't be used with Fargate tasks
2. Have you deployed an ECS service with these parameters? and does it actually work as expected? I am asking because of these comments: https://github.com/aws/aws-cdk/issues/18460#issuecomment-1014963300 and https://github.com/aws/aws-cdk/issues/18460#issuecomment-1015207388. If there is some actions users need to take for this to work after deployment, we need to document it. Or, hopefully do it for them in the cfn template. 
3. Please include an integration test.

File: packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts
Comment (flavioleggio) [reply packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts:108]: I added validation with positive integer values for maxSwap and values between 0 and 100 for swappiness. I also added a test to expect the constructor to throw with out of range values.

File: packages/@aws-cdk/aws-ecs/test/container-definition.test.ts
Comment (flavioleggio) [reply packages/@aws-cdk/aws-ecs/test/container-definition.test.ts:1776]: I added expected values in the templates for all the involved tests.

File: packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts
Comment (flavioleggio) [reply packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts:51]: I added some explicit descriptions for params meaning and accepted (valid) values. I also mentioned that these parameters are only supported with EC2 launch type, but I found no way to throw an error if the wrong launch type is supplied. Is there a clean and easy way to access the Service launchType attribute?

**Comment by flavioleggio** [comment]
Thanks for your review @madeline-k

> 2. Have you deployed an ECS service with these parameters? and does it actually work as expected? I am asking because of these comments: [#18460 (comment)](https://github.com/aws/aws-cdk/issues/18460#issuecomment-1014963300) and [#18460 (comment)](https://github.com/aws/aws-cdk/issues/18460#issuecomment-1015207388). If there is some actions users need to take for this to work after deployment, we need to document it. Or, hopefully do it for them in the cfn template.

I deployed a test stack and the parameters were correctly defined in the resulting task definition. However, I had no chance to check if swap memory was available inside the container and I did not figure out how, actually. Maybe I should ssh into the cluster instance and exec a `free` command on the running test container?

> 3. Please include an integration test.

I developed an integration test, but I have no permissions to create the VPC that I need for the ECS cluster in my company account, thus I cannot run the full integ test. What is the policy here? Should I commit the test without the corresponding `.expected.json` template? Wouldn't this make integ regression tests fail in CI?

**Comment by madeline-k** [comment]
> I developed an integration test, but I have no permissions to create the VPC that I need for the ECS cluster in my company account, thus I cannot run the full integ test. What is the policy here? Should I commit the test without the corresponding .expected.json template? Wouldn't this make integ regression tests fail in CI?

You're correct, you can't commit an integration test without the corresponding .expected.json template. Maybe I can help out here. Can you add the integration test to this PR? then I can run it and generate the expected.json template, and I'll push that to your branch.

**Comment by flavioleggio** [comment]
Ok, I’ll try to post it tomorrow

**Comment by flavioleggio** [comment]
```typescript
// packages/@aws-cdk/aws-ecs/test/ec2/integ.swap-parameters.ts

import * as ec2 from '@aws-cdk/aws-ec2';
import * as cdk from '@aws-cdk/core';
import * as ecs from '../../lib';
import { LinuxParameters } from '../../lib';

const app = new cdk.App();
const stack = new cdk.Stack(app, 'aws-ecs-integ');

const vpc = new ec2.Vpc(stack, 'Vpc', { maxAzs: 2 });

// ECS cluster to host EC2 task
const cluster = new ecs.Cluster(stack, 'EcsCluster', { vpc });
cluster.addCapacity('DefaultAutoScalingGroup', {
  instanceType: new ec2.InstanceType('t2.micro'),
});

// define task to run the container
const taskDefinition = new ecs.Ec2TaskDefinition(stack, 'TaskDefinition', {
  networkMode: ecs.NetworkMode.AWS_VPC,
});

// define linux parameters to enable swap
const linuxParameters = new LinuxParameters(stack, 'LinuxParameters', {
  maxSwap: 5e3,
  swappiness: 90,
});

// define container with linux parameters
new ecs.ContainerDefinition(stack, 'Container', {
  image: ecs.ContainerImage.fromRegistry('amazon/amazon-ecs-sample'),
  linuxParameters,
  memoryLimitMiB: 256,
  taskDefinition,
});

// define a service to run the task definition
new ecs.Ec2Service(stack, 'Service', {
  cluster,
  taskDefinition,
});

app.synth();
```

**Comment by TheRealAmazonKendra** [comment]
@flavioleggio Well, this is my bad. I didn't notice that the integration test was the old style and not the new. We'll need to update to the new style. I'll take care of that and run them for the expected output within the next week.

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

File: packages/@aws-cdk/aws-ecs/README.md
Comment (mrgrain) [inline packages/@aws-cdk/aws-ecs/README.md:501]:
```suggestion
    maxSwap: 5000,
```
I think this makes the example more readable (non blocking).

File: packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts
Comment (mrgrain) [inline packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts:109]: Could `maxSwap` be a Token? I believe we should check for this:
```suggestion
    if (!cdk.Token.isUnresolved(props.maxSwap) && props.maxSwap !== undefined && (!Number.isInteger(props.maxSwap) || props.maxSwap < 0)) {
```

File: packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts
Comment (mrgrain) [inline packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts:113]: Could `swappiness` be a Token? I believe we should check for this:
```suggestion
    if (!cdk.Token.isUnresolved(props.swappiness) && props.swappiness !== undefined && (!Number.isInteger(props.swappiness) || props.swappiness < 0 || props.swappiness > 100)) {
```

**Review by mrgrain — CHANGES_REQUESTED** [review]
@TheRealAmazonKendra Looks good now, I think we need the Token changes and then the PR is ready.

File: packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts
Comment (madeline-k) [inline packages/@aws-cdk/aws-ecs/lib/linux-parameters.ts:32]: This type should be `core.Size`. Sorry I didn't think of this in the first round of reviews!

**Review by madeline-k — CHANGES_REQUESTED** [review]
I only have one requested change here for this PR. 

Changing types is a breaking change, so we need that added before the PR is merged. 
In an ideal world, I think there would be more validation for this - error message when `maxSwap` is omitted, but `swappiness` is provided, and verify that these params can only be used when EC2 launch type is used. But we can add more validation in future PRs. I wouldn't hold this one for that.

**Comment by TheRealAmazonKendra** [comment]
@Mergifyio update

**Comment by TheRealAmazonKendra** [comment]
@mrgrain Can you take another look? I've re-updated the tests.

**Review by TheRealAmazonKendra — DISMISSED** [review]
I didn't notice the update in here to `sharedMemorySize`. We can't actually do that because it's a breaking change.

**Review by TheRealAmazonKendra — APPROVED** [review]
Looks great! Thanks for all your work on this!

**Pull request merged**
